File: server/packages/api/src/clients/s3.ts
import { S3Client } from "bun";
import { API_ENV_VAR } from "@/env/apiEnvVar";
import { getAPIEnvVar } from "@/env/getAPIEnvVar";

let s3Client: S3Client | null = null;

const buildVirtualHostedEndpoint = (
	endpoint: string,
	bucket: string,
): string => {
	const url = new URL(endpoint);
	if (!url.hostname.startsWith(`${bucket}.`)) {
		url.hostname = `${bucket}.${url.hostname}`;
	}

	return url.toString().replace(/\/$/, "");
};

export const getS3Client = (): S3Client => {
	if (!s3Client) {
		const S3_ACCESS_KEY_ID = getAPIEnvVar("S3_ACCESS_KEY_ID");
		const S3_SECRET_ACCESS_KEY = getAPIEnvVar("S3_SECRET_ACCESS_KEY");
		const S3_BUCKET = getAPIEnvVar("S3_BUCKET");
		const S3_ENDPOINT = getAPIEnvVar("S3_ENDPOINT");
		const S3_REGION = getAPIEnvVar("S3_REGION");
		const usePathStyle = process.env[API_ENV_VAR.S3_USE_PATH_STYLE] === "true";

		s3Client = new S3Client({
			accessKeyId: S3_ACCESS_KEY_ID,
			secretAccessKey: S3_SECRET_ACCESS_KEY,
			bucket: S3_BUCKET,
			endpoint: usePathStyle
				? S3_ENDPOINT
				: buildVirtualHostedEndpoint(S3_ENDPOINT, S3_BUCKET),
			region: S3_REGION,
			virtualHostedStyle: !usePathStyle,
		});
	}

	return s3Client;
};

File: models/agents/clients/s3.go
package clients

import (
	"context"
	"fmt"
	"io"
	"log"
	"os"
	"time"

	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

type S3Config struct {
	AccessKeyID     string
	AccessKeySecret string
	BucketName      string
	Endpoint        string
	Region          string
}

func LoadS3Config() S3Config {
	s3Config := S3Config{
		AccessKeyID:     os.Getenv("S3_ACCESS_KEY_ID"),
		AccessKeySecret: os.Getenv("S3_SECRET_ACCESS_KEY"),
		BucketName:      os.Getenv("S3_BUCKET"),
		Endpoint:        os.Getenv("S3_ENDPOINT"),
		Region:          os.Getenv("S3_REGION"),
	}

	if s3Config.AccessKeyID == "" || s3Config.AccessKeySecret == "" || s3Config.BucketName == "" || s3Config.Endpoint == "" || s3Config.Region == "" {
		log.Fatalf("incomplete S3 configuration")
	}

	return s3Config
}

type S3Client struct {
	s3Client *s3.Client
	config   S3Config
}

func NewS3Client(ctx context.Context) *S3Client {
	s3Config := LoadS3Config()

	cfg, err := config.LoadDefaultConfig(ctx,
		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(
			s3Config.AccessKeyID,
			s3Config.AccessKeySecret,
			"",
		)),
		config.WithRegion(s3Config.Region),
		config.WithBaseEndpoint(s3Config.Endpoint),
	)
	if err != nil {
		log.Fatalf("failed to load S3 config: %v", err)
	}

	s3Client := s3.NewFromConfig(cfg)

	return &S3Client{
		s3Client: s3Client,
		config:   s3Config,
	}
}

func (c *S3Client) Download(ctx context.Context, key string) ([]byte, error) {
	resp, err := c.s3Client.GetObject(ctx, &s3.GetObjectInput{
		Bucket: &c.config.BucketName,
		Key:    &key,
	})
	if err != nil {
		return nil, fmt.Errorf("GetObject %s: %w", key, err)
	}
	defer resp.Body.Close()

	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("reading %s: %w", key, err)
	}
	return bodyBytes, nil
}

func (c *S3Client) PresignDownload(ctx context.Context, key string, expiresIn time.Duration) (string, error) {
	if expiresIn <= 0 {
		expiresIn = 15 * time.Minute
	}

	presignClient := s3.NewPresignClient(c.s3Client)
	req, err := presignClient.PresignGetObject(ctx, &s3.GetObjectInput{
		Bucket: &c.config.BucketName,
		Key:    &key,
	}, func(opts *s3.PresignOptions) {
		opts.Expires = expiresIn
	})
	if err != nil {
		return "", fmt.Errorf("PresignGetObject %s: %w", key, err)
	}

	return req.URL, nil
}
